gettsim/benefits/alg2.py

```diff
@@ -221,58 +221,47 @@ def grossinc_alg2(household):
-    depending on monthly earnings."""
+def e_anr_frei(household, params):
+    """Determine the amount of income that is not deducted. Withdrawal rates
+    depend on monthly earnings and on the number of kids in the household.
+    § 30 SGB II. Since 01.04.2011 § 11b """
     # Calculate the amount of children below the age of 18.
-    child0_18_num = (household["child"] & household["age"].between(0, 18)).sum()
-
-    # 100€ is always 'free'
-    household.loc[(household["m_wage"] <= params["a2grf"]), "ekanrefrei"] = household[
-        "m_wage"
-    ]
-    # until 1000€, you may keep 20% (withdrawal rate: 80%)
-    household.loc[
-        (household["m_wage"].between(params["a2grf"], params["a2eg1"])), "ekanrefrei",
-    ] = params["a2grf"] + params["a2an1"] * (household["m_wage"] - params["a2grf"])
-    # from 1000 to 1200 €, you may keep only 10%
-    household.loc[
-        (household["m_wage"].between(params["a2eg1"], params["a2eg2"]))
-        & (child0_18_num == 0),
-        "ekanrefrei",
-    ] = (
-        params["a2grf"]
-        + params["a2an1"] * (params["a2eg1"] - params["a2grf"])
-        + params["a2an2"] * (household["m_wage"] - params["a2eg1"])
-    )
-    # If you have kids, this range goes until 1500 €,
-    household.loc[
-        (household["m_wage"].between(params["a2eg1"], params["a2eg3"]))
-        & (child0_18_num > 0),
-        "ekanrefrei",
-    ] = (
-        params["a2grf"]
-        + params["a2an1"] * (params["a2eg1"] - params["a2grf"])
-        + params["a2an2"] * (household["m_wage"] - params["a2eg1"])
-    )
-    # beyond 1200/1500€, you can't keep anything.
-    household.loc[
-        (household["m_wage"] > params["a2eg2"]) & (child0_18_num == 0), "ekanrefrei",
-    ] = (
-        params["a2grf"]
-        + params["a2an1"] * (params["a2eg1"] - params["a2grf"])
-        + params["a2an2"] * (params["a2eg2"] - params["a2eg1"])
-    )
-    household.loc[
-        (household["m_wage"] > params["a2eg3"]) & (child0_18_num > 0), "ekanrefrei",
-    ] = (
-        params["a2grf"]
-        + params["a2an1"] * (params["a2eg1"] - params["a2grf"])
-        + params["a2an2"] * (params["a2eg3"] - params["a2eg1"])
-    )
-    # Children income is fully deducted, except for the first 100 €.
-    household.loc[(household["child"]), "ekanrefrei"] = np.maximum(
-        0, household["m_wage"] - 100
+    num_childs_0_18 = (household["child"] & household["age"].between(0, 18)).sum()
+
+    if num_childs_0_18 == 0:
+        top_limit_2nd_interval = params["a2eg2"]
+    else:
+        top_limit_2nd_interval = params["a2eg3"]
+
+    cols = ["m_wage", "ekanrefrei"]
+    household.loc[:, cols] = household.groupby("pid")[cols].apply(
+        e_anr_frei_person, params, top_limit_2nd_interval
     )
 
     return household
+
+
+def e_anr_frei_person(person, params, top_limit_2nd_interval):
+    """Calculates the amount of income that is not deducted for each person."""
+    m_wage = person["m_wage"].iloc[0]
```

Comment:
Maybe we can avoid this chained indexing. What is ``m_wage``? Why is it save to take just the first entry for each person?

And even though this function is straightforward and beautifully structured, ``apply`` is notoriously slow. Is there an easy way to write this without the groupby and apply? I see no real aggregation.

Reply:
I also don't see the benefit here. The calculation is on the individual level anyway and makes distinctions based on individual and household characteristics. Why an additional loop?

Reply:
I did write an extra function on the individual level, because it allows to distinguish very clear the intervals of deduction. This is the only benefit. Otherwise we would slice the DataFrame everytime and then assign. Would you know a way, which combines readability and the fastness of slicing?

Reply:
person is always a one row Dataframe or a series. Therefore `person["m_wage"]`only contains one element.

Reply:
Some thoughts:

1. You cannot get away with the indentation on the right-hand-side of the equation, so we focus on the left-hand-side.
2. I would write each condition on a separate line like ``mask = ....`` and then ``df.loc[mask]``
3. ``child0_18_num > 0`` and ``child0_18_num == 0`` become ``has_minor/underage_children`` and ``~has_minor_children`` which is easier to read.
4. I am a little bit torn regarding this: ``household["m_wage"].between(params["a2eg1"], params["a2eg3"])``. On one hand it uses the designated pandas syntax, but on the other hand ``params["a2eg1"] <= household["m_wage"] <= params["a2eg3"]`` seems more readable to me in this instance.

Reply:
Don't worry much about speed here. [Premature optimization...](https://stackify.com/premature-optimization-evil/) :-). Will get there soon.

Reply:
Speed is one issue. But we have individual assignment within the household all the time. See calculation of taxable income, or wohngeld income test. Do we want to loop over `pid` there as well? And remove them when *actually* optimize?

Reply:
And yes, if we are already debating what "between" actually means, it is a useful thing to discard. In general, please avoid the less obvious Pandas constructs; remember we plan to also use CuDF and that does not implement the full set of Pandas methods.

Reply:
Sorry, but whats the conclusion now? Leave it as implemented or change to Tobias sugestion. Could you do a final word @hmgaudecker ?

Reply:
Leave it as is. We will have a more generic way for these linear spline-type functions eventually (they just appear everywhere), I just would not worry at this point.